File: src/generators/color.py
"""
Xxxx
xxxx
"""

import os
import re
import subprocess
import traceback

import numpy as np
from PIL import Image, ImageFilter
from sklearn.cluster import KMeans


def generate_color_svg(
    input_path,
    output_path,
    num_colors=32,  # Colores base más realista
    # color_tolerance=25,  # Tolerancia más fina
    # min_area=10,  # Área mínima de región
    turdsize=2,  # Detalles finos
    blur_radius=1,  # Suavizado previo
):
    """
    Genera SVG de alta calidad desde PNG con alpha, preservando colores y formas.
    """
    if os.path.exists(output_path):
        return

    try:
        # --- Cargar y preparar imagen ---
        img = Image.open(input_path).convert("RGBA")
        width, height = img.size
        rgba = np.array(img)
        rgb_arr = rgba[..., :3]
        alpha_arr = rgba[..., 3]

        # Crear máscara de píxeles visibles
        visible_mask = alpha_arr > 20

        # --- Detectar si es B/N ---
        if visible_mask.sum() > 0:
            visible_pixels = rgb_arr[visible_mask]
            saturation = np.std(visible_pixels, axis=1).mean()
            is_bw = saturation < 15
        else:
            is_bw = True

        svg_layers = []

        if is_bw:
            # --- Modo Blanco y Negro ---
            gray = np.array(img.convert("L"))
            # Umbralización adaptativa
            threshold = np.median(gray[visible_mask]) if visible_mask.sum() > 0 else 128
            mask_arr = np.where((gray < threshold) & visible_mask, 0, 255).astype(
                np.uint8
            )

            mask = Image.fromarray(mask_arr, mode="L")

            # Suavizado opcional
            if blur_radius > 0:
                mask = mask.filter(ImageFilter.GaussianBlur(blur_radius))

            t_bmp = f"{output_path}.temp.bmp"
            t_svg = f"{output_path}.temp.svg"

            mask.save(t_bmp)

            subprocess.run(
                [
                    "potrace",
                    t_bmp,
                    "-s",
                    "-o",
                    t_svg,
                    "--flat",
                    "--turdsize",
                    str(turdsize),
                    "--alphamax",
                    "0.5",  # Suavizar curvas
                ],
                check=True,
                capture_output=True,
            )

            with open(t_svg, "r", encoding="utf-8") as f:
                content = f.read()

            path_match = re.search(r'<path d="([^"]+)"', content)
            transform_match = re.search(r'<g transform="([^"]+)"', content)

            if path_match:
                d = path_match.group(1)
                transform = transform_match.group(1) if transform_match else ""
                svg_layers.append(
                    f'<g transform="{transform}"><path d="{d}" fill="#000000" stroke="none" /></g>'
                )

            os.remove(t_bmp)
            os.remove(t_svg)

        else:
            # --- Modo Color: Clustering K-means mejorado ---
            visible_pixels = rgb_arr[visible_mask]

            # Clustering con K-means para colores reales
            kmeans = KMeans(
                n_clusters=min(num_colors, len(visible_pixels)),
                random_state=42,
                n_init=10,
            )
            kmeans.fit(visible_pixels)

            # Obtener colores centrales y contar píxeles
            colors = kmeans.cluster_centers_.astype(int)
            labels = kmeans.labels_

            # Mapear labels de vuelta a la imagen completa
            full_labels = np.full((height, width), -1, dtype=int)
            full_labels[visible_mask] = labels

            # Calcular área de cada color
            unique_labels, counts = np.unique(labels, return_counts=True)

            # Ordenar por área (colores más grandes primero, como fondo)
            sorted_indices = np.argsort(-counts)

            for idx in sorted_indices:
                label = unique_labels[idx]
                color = tuple(colors[label])

                # Filtrar colores casi blancos (fondo)
                if sum(color) > 740:
                    continue

                # Crear máscara para este color
                color_mask = full_labels == label
                mask_arr = np.where(color_mask, 0, 255).astype(np.uint8)

                mask = Image.fromarray(mask_arr, mode="L")

                # Cerrar pequeños huecos
                mask = mask.filter(ImageFilter.MinFilter(3))
                mask = mask.filter(ImageFilter.MaxFilter(3))

                # Suavizado
                if blur_radius > 0:
                    mask = mask.filter(ImageFilter.GaussianBlur(blur_radius))

                t_bmp = f"{output_path}.layer{label}.bmp"
                t_svg = f"{output_path}.layer{label}.svg"

                mask.save(t_bmp)

                subprocess.run(
                    [
                        "potrace",
                        t_bmp,
                        "-s",
                        "-o",
                        t_svg,
                        "--flat",
                        "--turdsize",
                        str(turdsize),
                        "--alphamax",
                        "0.8",  # Curvas más suaves para color
                        "--opttolerance",
                        "0.2",
                    ],
                    check=True,
                    capture_output=True,
                )

                with open(t_svg, "r", encoding="utf-8") as f:
                    content = f.read()

                path_match = re.search(r'<path d="([^"]+)"', content)
                transform_match = re.search(r'<g transform="([^"]+)"', content)

                if path_match:
                    d = path_match.group(1)
                    transform = transform_match.group(1) if transform_match else ""
                    hex_color = f"#{color[0]:02x}{color[1]:02x}{color[2]:02x}"
                    svg_layers.append(
                        f'<g transform="{transform}">'
                        f'<path d="{d}" fill="{hex_color}" stroke="none" />'
                        "</g>"
                    )

                os.remove(t_bmp)
                os.remove(t_svg)

        # --- Guardar SVG final ---
        with open(output_path, "w", encoding="utf-8") as f:
            f.write('<?xml version="1.0" encoding="UTF-8" standalone="no"?>\n')
            f.write(
                f'<svg version="1.1" xmlns="http://www.w3.org/2000/svg" '
                f'width="{width}" height="{height}" '
                f'viewBox="0 0 {width} {height}">\n'
            )
            for layer in svg_layers:
                f.write(f"  {layer}\n")
            f.write("</svg>\n")

        print(
            f"🎨 SVG generado: {os.path.basename(output_path)} ({len(svg_layers)} capas)"
        )

    except Exception as e:
        print(f"❌ Error: {os.path.basename(input_path)}: {e}")
        traceback.print_exc()


# === EJEMPLO DE USO ===
if __name__ == "__main__":
    # Para logos con pocos colores
    generate_color_svg("logo.png", "logo.svg", num_colors=16, blur_radius=0.5)

    # Para ilustraciones complejas
    generate_color_svg(
        "illustration.png", "illustration.svg", num_colors=48, blur_radius=1
    )

File: rust/src/generators/color.rs
use image::{DynamicImage, Luma};
use std::path::Path;
use std::process::Command;
use std::fs;
use anyhow::{Result, anyhow};
use tempfile::NamedTempFile;
use kmeans_colors::get_kmeans;
use palette::{Srgb, Lab, FromColor, IntoColor};

pub fn generate_color_svg(img: &DynamicImage, output_path: &Path, num_colors: u32) -> Result<()> {
    if output_path.exists() {
        return Ok(());
    }

    let rgba = img.to_rgba8();
    let (width, height) = rgba.dimensions();
    
    // Collect visible pixels for K-means
    let mut pixels = Vec::new();
    for pixel in rgba.pixels() {
        if pixel.0[3] > 20 {
            let srgb = Srgb::new(
                pixel.0[0] as f32 / 255.0,
                pixel.0[1] as f32 / 255.0,
                pixel.0[2] as f32 / 255.0,
            );
            let lab: Lab = srgb.into_color();
            pixels.push(lab);
        }
    }

    if pixels.is_empty() {
        return Err(anyhow!("No visible pixels found for color quantization"));
    }

    let k = num_colors.min(pixels.len() as u32) as usize;
    let result = get_kmeans(k, 10, 0.005, false, &pixels, 12345);
    let colors = result.centroids;
    
    let mut svg_layers = Vec::new();

    for (i, Lab { l, a: ca, b: cb, .. }) in colors.iter().enumerate() {
        let srgb: Srgb = Srgb::from_color(Lab::new(*l, *ca, *cb));
        let r_u8 = (srgb.red * 255.0) as u8;
        let g_u8 = (srgb.green * 255.0) as u8;
        let b_u8 = (srgb.blue * 255.0) as u8;

        if r_u8 > 245 && g_u8 > 245 && b_u8 > 245 { continue; } // Skip background

        let mut mask = image::ImageBuffer::new(width, height);
        let mut found = false;
        
        for (x, y, pixel) in rgba.enumerate_pixels() {
            if pixel.0[3] > 20 {
                let px_srgb = Srgb::new(
                    pixel.0[0] as f32 / 255.0,
                    pixel.0[1] as f32 / 255.0,
                    pixel.0[2] as f32 / 255.0,
                );
                let px_lab: Lab = px_srgb.into_color();
                
                // Find nearest centroid
                let mut min_dist = f32::MAX;
                let mut best_idx = 0;
                for (idx, centroid) in colors.iter().enumerate() {
                    let d = (px_lab.l - centroid.l).powi(2) + (px_lab.a - centroid.a).powi(2) + (px_lab.b - centroid.b).powi(2);
                    if d < min_dist {
                        min_dist = d;
                        best_idx = idx;
                    }
                }

                if best_idx == i {
                    mask.put_pixel(x, y, Luma([0u8]));
                    found = true;
                } else {
                    mask.put_pixel(x, y, Luma([255u8]));
                }
            } else {
                mask.put_pixel(x, y, Luma([255u8]));
            }
        }

        if !found { continue; }

        let temp_bmp = NamedTempFile::new_in(".")?;
        let bmp_path = temp_bmp.path().with_extension("bmp");
        mask.save(&bmp_path)?;

        let temp_svg = NamedTempFile::new_in(".")?;
        let svg_tmp_path = temp_svg.path().with_extension("svg");

        let status = Command::new("potrace")
            .args(&[
                bmp_path.to_str().unwrap(),
                "-s",
                "-o",
                svg_tmp_path.to_str().unwrap(),
                "--flat",
                "--turdsize", "2",
                "--alphamax", "0.8",
            ])
            .status()?;

        if status.success() {
            let content = fs::read_to_string(&svg_tmp_path)?;
            let hex_color = format!("#{:02x}{:02x}{:02x}", r_u8, g_u8, b_u8);
            
            // Robustly extract the content between <svg ...> and </svg>
            if let Some(start_idx) = content.find("<svg") {
                if let Some(content_start) = content[start_idx..].find('>') {
                    let inner_content_start = start_idx + content_start + 1;
                    if let Some(end_idx) = content.rfind("</svg>") {
                        let inner_content = &content[inner_content_start..end_idx];
                        // Replace common black fill values
                        let colored_content = inner_content
                            .replace("fill=\"black\"", &format!("fill=\"{}\"", hex_color))
                            .replace("fill=\"#000000\"", &format!("fill=\"{}\"", hex_color));
                        svg_layers.push(colored_content);
                    }
                }
            }
        }
        
        let _ = fs::remove_file(bmp_path);
        let _ = fs::remove_file(svg_tmp_path);
    }

    let mut final_svg = format!(
        "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n\
        <svg version=\"1.1\" xmlns=\"http://www.w3.org/2000/svg\" \
        width=\"{}\" height=\"{}\" viewBox=\"0 0 {} {}\">\n",
        width, height, width, height
    );

    for layer in svg_layers {
        final_svg.push_str(&layer);
        final_svg.push('\n');
    }
    final_svg.push_str("</svg>");

    fs::write(output_path, final_svg)?;
    println!("🎨 SVG Color OK: {:?}", output_path.file_name().unwrap());
    Ok(())
}
